### PR TITLE
Make component instantiation more general in order to support React 16

### PR DESCRIPTION
**Pull request by dpoindexter**
omniscient components have to behave like React class constructors and like functions that return React elements. Before React 16, an internal implementation detail of React's component instantiation allowed omniscient to detect whether to return a React component instance or a React element.

This PR makes that detection logic more general. It still relies on React's implementation (using the `new` keyword to instantiate component instances) but that seems unlikely to change. The caveat is that an end user can't call `new Component(props);` directly, but that doesn't seem like it was intended to be a supported use regardless.

Without this change, the omniscient would basically create an addition stateless functional component wrapper around every omniscient component. This actually still works fine in most cases, but breaks features that are intended for React class components only, like `childContextTypes`.

With this change, we have omniscient running against React 16 on a pretty large component library, so it looks like it might be ready for a new version release!

### CODE REVIEW

File: component.js [hunk 1]
```diff
@@ -257,15 +265,21 @@ function factory (initialOptions) {
        * @returns {ReactElement}
        * @api public
        */
-      var create = function (keyOrProps, propsOrPublicContext, ReactUpdateQueue) {
-        // After stateless arrow functions was allowed as components, react will instantiate
-        // the `create` function if it has a prototype. We are passed `props`, `publicContext`
-        // and `ReactUpdateQueue`.
-        // https://github.com/facebook/react/blob/88bae3fb73511893519195e451c56896463f669b/src/renderers/shared/reconciler/ReactCompositeComponent.js#L154-L171
-        if (typeof ReactUpdateQueue == 'object' && !_isNode(ReactUpdateQueue)) {
+      var create = function create (keyOrProps, propsOrPublicContext, updater) {
```
Comment (torgeir) [inline component.js]: The function now being named, we might as well make it work without `var create`?

Its used a bit further down for the following, but I see no real use for the assignment..

```js
  if (methodStatics) {
    create = assign(create, methodStatics);
  }
```

**Comment by mikaelbr** [comment]
Great! 🎉 Don't really need to provide the dist files here. They are only making the commits "noisy"

**Comment by dpoindexter** [comment]
A couple of notes:

> I would love a test or two to lock down the need for isReactComponent on the prototype chain and to show the breakage for features that are intended for React class components only.

I took a stab at this, but the test is pretty indirect. It spies on `console.error` to see if React logs anything about `childContextTypes`. I couldn't figure out a more obvious way to test this without introducing a whole new system (i.e. enzyme), but I'm open to suggestions. You can see the test fail if you comment out `component.js:330`.

I also added prettier as a dev dependency. Otherwise contributors might accidentally, say, run a globally installed, older version of prettier that doesn't understand the newer prettier configuration. 😃

**Comment by torgeir** [comment]
Yeah I agree. So indirect I have second thoughts if we should bother with it, especially asserting on the error message.. How about something like this instead?

```js
    test('exposes `isReactComponent` on the prototype, not to create additional element wrapper when rendered by React', function() {
      var Component = component(() => React.createElement('div'));
      const element = React.createElement(Component);
      expect(element.type.prototype.isReactComponent).toBeDefined();
    });
```

> I also added prettier as a dev dependency

Nice!

Do you mind squashing up the commits? 

Am I right to assume the `ReactUpdateQueue`/`this.constructor` change only works with React >= 16? If so we need to be bumping the peer dep as well. (And probably a patch release for those wanting to stay on 4.2 limiting it to < 16 @mikaelbr)

Other than that I think this is starting to look good!

**Comment by dpoindexter** [comment]
@torgeir, re: test suggestion and squashing, done and done.

I don't think the version range needs to be updated -- this should be fully backwards-compatible. The Travis test suite is running against React 15, if I'm not mistaken.

File: package.json [hunk 2]
```diff
@@ -25,6 +25,8 @@
     "immutable": "^3.8.1",
     "jest": "^21.2.1",
     "lint-staged": "^4.2.3",
+    "prettier": "^1.7.4",
+    "prop-types": "^15.6.0",
```
Comment (mikaelbr) [inline package.json]: Do we use prop-types anywhere?

File: __tests__/component.test.js [hunk 3]
```diff
@@ -1,6 +1,7 @@
 var Immutable = require('immutable');
 var Cursor = require('immutable/contrib/cursor');
 var createClass = require('create-react-class');
+var PropTypes = require('prop-types');
```
Comment (mikaelbr) [inline __tests__/component.test.js]: Is this just imported but not used?

**Comment by mikaelbr** [comment]
I changed the travis tests some time ago. It now uses 16 (I think that is more correct and relevant). I couldn't find a good way to test with multiple dependencies.

File: component.js [hunk 4]
```diff
@@ -261,19 +269,21 @@ function factory(initialOptions) {
...
-        propsOrPublicContext,
-        ReactUpdateQueue
-      ) {
-        // After stateless arrow functions was allowed as components, react will instantiate
-        // the `create` function if it has a prototype. We are passed `props`, `publicContext`
-        // and `ReactUpdateQueue`.
-        // https://github.com/facebook/react/blob/88bae3fb73511893519195e451c56896463f669b/src/renderers/shared/reconciler/ReactCompositeComponent.js#L154-L171
-        if (typeof ReactUpdateQueue == 'object' && !_isNode(ReactUpdateQueue)) {
+      function create(keyOrProps, propsOrPublicContext, updater) {
+        // `create` must handle two scenarios (given a component like `var MyComponent = component(() => <div />)`)
```
Comment (mikaelbr) [inline component.js]: Great documentation! 🎉

**Comment by dpoindexter** [comment]
@mikaelbr `prop-types` was left over from my first attempt to write a test for the component instantiation stuff using `childContextTypes`. I've removed it.

It looks like it might be possible to test multiple versions of packages with something like `[test-all-versions](https://www.npmjs.com/package/test-all-versions)`. I ran the test suite locally against React 15.x and everything passes -- I'm pretty sure this is a backwards compatible change, since older versions of React also instantiate instance classes with the `new` keyword.

**Comment by mikaelbr** [comment]
Cool. I'll investigate TAV some more. Looks like exactly what we need. Great tip.

**Comment by dpoindexter** [comment]
Anything else missing for a release? Happy to work on it if so.

**Comment by mikaelbr** [comment]
I think we're good. I can add TAV on a later point (doesn't block this release).

**Pull request merged**